### PR TITLE
8255765: Shenandoah: Isolate concurrent, degenerated and full GC

### PR DESCRIPTION
**Pull request by zhengyu123**
The purpose of this patch is to isolate concurrent, degenerated and full gc implementation, so that, makes each GC implementation more straightforward and clean, and improves readability.

Current implementation emphasis code sharing, e.g. degenerated GC reuses concurrent GC's ops. It was not a problem in the beginning, when they actually behave similarly. Since concurrent GC moved root processing into concurrent phases, code started to diverge, we started to put bandages to make the shared ops work for both concurrent and degenerated GC, that made code hard to read and error prone.

The patch breaks up GCs into 3 (mainly just concurrent and degenerated GC, as full GC already standalone) easy to identify and understand classes (ShenandoahConcurrentGC, ShenandoahDegeneratedGC and ShenandoahMarkCompactGC), subclasses of ShenandoahGC class.

The three GCs still keep vmop/entry/op paradigm, but encapsulate GC control flow inside their own classes, as ShenandoahMarkCompact GC already does. So that, concurrent and degenerated GC no longer share ops and op implementations no longer need to consider other GC modes, which results in simplifying implementation and improving readability. Code sharing is achieved via helper methods provided by ShenandoahHeap.

Test:
 - [x] hotspot_gc_shenandoah
 - [x] nightly tests

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255765](https://bugs.openjdk.java.net/browse/JDK-8255765): Shenandoah: Isolate concurrent, degenerated and full GC

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1964/head:pull/1964`
`$ git checkout pull/1964`

### CODE REVIEW

**Review by rkennke — COMMENTED** [review]
Nice! I really like how that moves some burden out from ShHeap (and ShControlThread) to more apprppriate places. Just the following two remarks.

**Comment by rkennke** [comment]
@shipilev should also look at this.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp [hunk 1]
```diff
@@ -362,85 +337,38 @@ class ShenandoahHeap : public CollectedHeap {
...
-  const char* init_mark_event_message() const;
-  const char* final_mark_event_message() const;
-  const char* conc_mark_event_message() const;
-  const char* degen_event_message(ShenandoahDegenPoint point) const;
-
-// Helpers
-  void finish_mark();
-  void prepare_evacuation();
+public:
+  void notify_gc_progress()    { _progress_last_gc.set();}
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:369]: Align the closing brace with the next line?

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.cpp [hunk 2]
```diff
@@ -193,7 +193,6 @@ void ShenandoahConcurrentRootScanner::roots_do(OopClosure* oops, uint worker_id)
   _vm_roots.oops_do(oops, worker_id);
 
   if (!heap->unload_classes()) {
-    AlwaysTrueClosure always_true;
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.cpp:196]: This change seems unrelated?

**Review by shipilev — CHANGES_REQUESTED** [review]
Some minor nits from the initial review. Please make sure `tier1` and `tier2` with Shenandoah still pass.

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.cpp (hunk 2 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.cpp:196]: Right. Restored and will file a separate bug to clean it up.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+  static const char* msg = "Concurrent cleanup";
+  ShenandoahConcurrentPhase gc_phase(msg, ShenandoahPhaseTimings::conc_cleanup_complete, true /* log_heap_usage */);
+  EventMark em("%s", msg);
+
+  // This phase does not use workers, no need for setup
+  heap->try_inject_alloc_failure();
+  op_cleanup_complete();
+}
+
+// Actual work for the phases
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:412]: I think we can drop this comment.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+    _nmethod_itr(ShenandoahCodeRoots::table()),
+    _dedup_roots(phase),
+    _phase(phase),
+    _concurrent_class_unloading(ShenandoahConcurrentRoots::should_do_concurrent_class_unloading()) {
+    if (_concurrent_class_unloading) {
+      MutexLocker mu(CodeCache_lock, Mutex::_no_safepoint_check_flag);
+      _nmethod_itr.nmethods_do_begin();
+    }
+
+     _dedup_roots.prologue();
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:659]: It looks to me this line in misindented: one stray space?

File: src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+      // Note that we can only do this for "outside-cycle" degens, otherwise we would risk
+      // changing the cycle parameters mid-cycle during concurrent -> degenerated handover.
+      heap->set_unload_classes(heap->heuristics()->can_unload_classes());
+
+      op_reset();
+
+      // STW mark
+      op_mark();
+    case _degenerated_mark:
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp:115]: New line before this `case`?

File: src/hotspot/share/gc/shenandoah/shenandoahGC.hpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  };
+
+  virtual bool collect(GCCause::Cause cause) = 0;
+  static const char* degen_point_to_string(ShenandoahDegenPoint point);
+
+protected:
+  static void update_roots(bool full_gc);
+};
+
+#endif
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahGC.hpp:66]: Should be `#endif // SHARE_GC_SHENANDOAH_SHENANDOAHGC_HPP`?

**Review by shipilev — APPROVED** [review]
I have only minor nits left.

**Review by rkennke — APPROVED** [review]
Looks good!

**Review by shipilev — APPROVED** [review]
Looks good.
